timemachine/cpp/src/rmsd_restraint.cu

```diff
@@ -74,87 +102,82 @@ void RMSDRestraint<RealType>::execute_device(
     unsigned long long *d_u,  // buffered
     cudaStream_t stream) {
 
-
     if(N != N_) {
         throw std::runtime_error("N_! = N");
     }
-
-    std::vector<double> h_x(N*3);
-
-    gpuErrchk(cudaMemcpy(&h_x[0], d_x, h_x.size()*sizeof(*d_x), cudaMemcpyDeviceToHost));
-
+    const int tpb = 32;
     const int B = h_atom_map_.size() / 2;
 
-    Eigen::MatrixXd x1(B,3);
-    Eigen::MatrixXd x2(B,3);
+    dim3 grid_blocks(B, 3);
 
-    // compute centroids etc. on the GPU directly
+    gpuErrchk(cudaMemset(d_centroid_a_, 0.0, B*3*sizeof(RealType)));
+    gpuErrchk(cudaMemset(d_centroid_b_, 0.0, B*3*sizeof(RealType)));
 
-    for(int b=0; b < B; b++) {
-
-        int src_idx = h_atom_map_[b*2+0];
-        x1(b, 0) = h_x[src_idx*3+0];
-        x1(b, 1) = h_x[src_idx*3+1];
-        x1(b, 2) = h_x[src_idx*3+2];
+    k_centroid_sums<RealType><<<grid_blocks, tpb, 0, stream>>>(
+        d_atom_map_,
+        B,
+        d_x,
+        d_centroid_a_,
+        d_centroid_b_
+    );
+    gpuErrchk(cudaPeekAtLastError());
 
-        int dst_idx = h_atom_map_[b*2+1];
-        x2(b, 0) = h_x[dst_idx*3+0];
-        x2(b, 1) = h_x[dst_idx*3+1];
-        x2(b, 2) = h_x[dst_idx*3+2];
-    }
+    // Need to use RowMajor storage to allow copying
+    Eigen::Matrix<RealType, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor> x1(B, 3);
+    Eigen::Matrix<RealType, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor> x2(B, 3);
 
+    gpuErrchk(cudaMemcpy(x1.data(), d_centroid_a_, B*3*sizeof(RealType), cudaMemcpyDeviceToHost));
+    gpuErrchk(cudaMemcpy(x2.data(), d_centroid_b_, B*3*sizeof(RealType), cudaMemcpyDeviceToHost));
 
-    Eigen::Vector3d h_a_centroid = x1.colwise().mean();
-    Eigen::Vector3d h_b_centroid = x2.colwise().mean();
+    Eigen::Matrix<RealType, 3, 1> h_a_centroid_mean = x1.colwise().mean();
+    Eigen::Matrix<RealType, 3, 1> h_b_centroid_mean = x2.colwise().mean();
 
-    x1 = x1.rowwise() - h_a_centroid.transpose();
-    x2 = x2.rowwise() - h_b_centroid.transpose();
+    x1 = x1.rowwise() - h_a_centroid_mean.transpose();
+    x2 = x2.rowwise() - h_b_centroid_mean.transpose();
 
-    Eigen::MatrixXd c = x2.transpose() * x1;
+    // Keep Matrices in doubles to avoid bad behavior in float32
+    Eigen::Matrix<double, 3, 3> c = (x2.transpose().template cast<double> () * x1.template cast<double> ());
 
-    Eigen::JacobiSVD<Eigen::MatrixXd> svd(c, Eigen::ComputeFullU | Eigen::ComputeFullV);
+    Eigen::JacobiSVD<Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic>> svd(c, Eigen::ComputeFullU | Eigen::ComputeFullV);
```

Comment:
If the following matrix operations are done using float32, things misbehave. Everything up to this point is numerically the same, but afterwards it blows up.

Reply:
Thats... concerning. Is it just the SVD itself that needs double? or everything afterwards? does doing a double precision SVD and single precision backprop work?

Reply:
Will dig into it and find where it breaks, the eigen templating was giving me a hard time

Reply:
It was the SVD, everything else works as it should.

Reply:
I had thought i had nailed it down, but looked like everything past the SVD needs to be double. Wasn't able to nail down what line was spitting out bad numbers.